### PR TITLE
added GPTNeoForTokenClassification

### PR DESCRIPTION
**Pull request by peter-sk**
# What does this PR do?

It adds the class GPTNeoForTokenClassification, which allows using GPT Neo models for token classification tasks. The implementation follows the one for other models (such as GPT2) closely and simply adds a linear layer after the hidden states.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker
@younesbelkada

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Could you make sure the CI tests are green? Can reveiw then!

**Comment by peter-sk** [comment]
@ArthurZucker
Sure. I'm getting the hang of it. Now, the only failing tests are connected to flax and seem unrelated to this pull request.

**Comment by peter-sk** [comment]
If the flax errors are not due to the PR, this is ready to be reviewed, @ArthurZucker and @younesbelkada :-)

**Comment by peter-sk** [comment]
I just checked the logs for the remaining errors one more time. The errors are related to the import of the optax library, where jax.Array is used in a type. Apparently there is no name "Array" in the top-level namespace of the jax module.

I cannot see how this could be related to my PR.

**Comment by peter-sk** [comment]
The jax version used in the examples_flax test is 0.3.6:
Collecting jax!=0.3.2,<=0.3.6,>=0.2.8 (from transformers==4.28.0.dev0)
    Using cached jax-0.3.6-py3-none-any.whl
This version clearly has no Array class.
I am unsure why such an old version should be used?

**Comment by peter-sk** [comment]
Figured out that optax <= 0.1.4 is needed. And found out that upstream/main has that change already 👍  Now everything should be cleared for review.

**Comment by ArthurZucker** [comment]
Cool! Reviewing now

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 1]
```diff
@@ -926,3 +927,101 @@ def forward(
...
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(GPT_NEO_INPUTS_DOCSTRING)
+    # fmt: off
+    @add_code_sample_docstrings(
+        checkpoint="EleutherAI/gpt-neo-125m",
+        output_type=TokenClassifierOutput,
+        config_class=_CONFIG_FOR_DOC,
+        expected_loss=0.25,
+        expected_output=["Lead", "Lead", "Lead", "Position", "Lead", "Lead", "Lead", "Lead", "Lead", "Lead", "Lead", "Lead"],
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:968]: also let's remove the fmt on/off here

**Review by ArthurZucker — APPROVED** [review]
Thanks for working on this! Left a few comments but looks good otherwise!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your work on this, I have the same comments as @ArthurZucker !

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 2]
```diff
@@ -926,3 +927,101 @@ def forward(
...
+        super().__init__(config)
+        self.num_labels = config.num_labels
+
+        self.transformer = GPTNeoModel(config)
+        if hasattr(config, "classifier_dropout") and config.classifier_dropout is not None:
+            classifier_dropout = config.classifier_dropout
+        elif hasattr(config, "hidden_dropout") and config.hidden_dropout is not None:
+            classifier_dropout = config.hidden_dropout
+        else:
+            classifier_dropout = 0.1
```
[the remark this answers is not in the document]
Comment (peter-sk) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:950]: Done.

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 3]
```diff
@@ -926,3 +927,101 @@ def forward(
...
+        self.dropout = nn.Dropout(classifier_dropout)
+        self.classifier = nn.Linear(config.hidden_size, config.num_labels)
+
+        # Model parallel
+        self.model_parallel = False
+        self.device_map = None
```
[the remark this answers is not in the document]
Comment (peter-sk) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py]: Done.

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (peter-sk) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py]: Done.
Done.

**Comment by peter-sk** [comment]
All done and ready to be merged, @ArthurZucker and @younesbelkada 👍

**Comment by peter-sk** [comment]
I implemented the same change as for GPTNeoXForTokenClassification, i.e., I removed the hasattr etc. and just use config.classifier_dropout directly.

**Review by younesbelkada — APPROVED** [review]
Thank you very much again!

**Review by sgugger — APPROVED** [review]
Thanks a lot! Can you just solve the conflicts so we can merge the PR?

**Pull request merged**